syntax = "proto3";

package com.whatsapp.proto;

message Message {
    SenderKeyDistributionMessage sender_key_distribution_message   = 2;
    oneof body {
        string          conversation        = 1;
        ImageMessage    image_message       = 3;
        ContactMessage  contact_message     = 4;
        LocationMessage location_message    = 5;
        ExtendedTextMessage  extended_text_message = 6;
        DocumentMessage document_message    = 7;
        AudioMessage    audio_message       = 8;
        VideoMessage    video_message       = 9;
        VideoMessage    video_message       = 9;
        VideoMessage    video_message       = 9;
        VideoMessage    video_message       = 9;
    }
}

message SenderKeyDistributionMessage {
    string groupId = 1;
    bytes axolotl_sender_key_distribution_message = 2;
}

message ImageMessage {
    bytes url             = 1;
    string mime_type      = 2;
    string caption        = 3;
    bytes file_sha256     = 4;
    uint64 file_length    = 5;
    uint32 height         = 6;
    uint32 width          = 7;
    bytes media_key       = 8;
    bytes file_enc_sha256 = 9;
    bytes jpeg_thumbnail  = 16;
}

message AudioMessage {
    bytes url             = 1;
    string mime_type      = 2;
    bytes file_sha256     = 3;
    uint64 file_length    = 4;
    uint64 duration       = 5;
    uint32  unk           = 6;
    bytes media_key       = 7;
}

message VideoMessage {
    bytes url             = 1;
    string mime_type      = 2;
    bytes file_sha256     = 3;
    uint64 file_length    = 4;
    uint64 seconds        = 5;
    bytes media_key       = 6;
    uint32 height         = 9;
    uint32 width          = 10;
    bytes file_enc_sha256 = 11;
    bytes jpeg_thumbnail  = 16;
    ContextInfo context_info = 17;
    bytes streaming_sidecar = 18;
}

message LocationMessage {
    double degrees_latitude   = 1;
    double degrees_longitude  = 2;
    string name               = 3;
    string address            = 4;
    string url                = 5;
    bytes  jpeg_thumbnail     = 16;

}
message DocumentMessage {
    string url            = 1;
    string mimeType       = 2;
    string title          = 3;
    bytes file_sha256     = 4;
    uint64 file_length    = 5;
    uint32 page_count     = 6;
    bytes media_key       = 7;
    bytes jpeg_thumbnail  = 16;
}


message ExtendedTextMessage {
    string text              = 1;
    string matched_text      = 2;
    string canonical_url     = 4;
    string description       = 5;
    string title             = 6;
    bytes jpeg_thumbnail     = 16;
    ContextInfo context_info = 17;
}

message ContextInfo {
    string stanza_id = 1;
    string participant = 2;
    Message quoted_message_ = 3;
    string remote_jid = 4;
    string mentioned_jid = 15;
}

message ContactMessage {
    string display_name = 1;
    string vcard = 16;
}
